src/Analyser/TypeSpecifier.php

```diff
@@ -1005,6 +1005,16 @@ private function specifyTypesForConstantBinaryExpression(
 						if ($isSize->no()) {
 							continue;
 						}
+
+						if ($innerType->isList()->yes()) {
+							// turn optional offsets non-optional
+							$valueTypesBuilder = ConstantArrayTypeBuilder::createEmpty();
+							for ($i = 0; $i < $constantType->getValue(); $i++) {
+								$offsetType = new ConstantIntegerType($i);
+								$valueTypesBuilder->setOffsetValueType($offsetType, $innerType->getOffsetValueType($offsetType));
```

Comment:
What about doing intersect with HasOffsetType instead of the for loop?

Reply:
I tried using `HasOffsetValueType` but it still requires the loop

```diff
-                                               $valueTypesBuilder = ConstantArrayTypeBuilder::createEmpty();
+                                               $hasOffsetTypes = [];
                                                for ($i = 0; $i < $constantType->getValue(); $i++) {
-                                                       $offsetType = new ConstantIntegerType($i);
-                                                       $valueTypesBuilder->setOffsetValueType($offsetType, $argType->getOffsetValueType($offsetType));
+                                                       $hasOffsetTypes[] = new HasOffsetValueType(new ConstantIntegerType($i), $argType->getOffsetValueType(new ConstantIntegerType(
$i)));
                                                }
-                                               $valueTypes = $this->create($exprNode->getArgs()[0]->value, $valueTypesBuilder->getArray(), $context, false, $scope, $rootExpr);
+                                               $valueTypes = $this->create($exprNode->getArgs()[0]->value, TypeCombinator::intersect($argType, ...$hasOffsetTypes), $context, false,
 $scope, $rootExpr);
```

and it leas to less precise types, I think

```
--- Expected
+++ Actual
@@ @@
-'array{ListCount\A, ListCount\A, ListCount\A}'
+'list<ListCount\A>&hasOffsetValue(0, ListCount\A)&hasOffsetValue(1, ListCount\A)&hasOffsetValue(2, ListCount\A)'
```

Reply:
to drop the loop we could maybe use `new HasOffsetValueType(IntegerRangeType::fromInterval(0, $constantType->getValue(), $offsetValueType)` but this doesn't work because $offsetValueType is different per offset